### PR TITLE
Num Bucket for GConstruct

### PR DESCRIPTION
**Pull request by jalencato**
*Issue #, if available:*

*Description of changes:*

This PR contains modification about test case in gsprocessing, which will delete after develop on this feature.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gconstruct/transform.py [hunk 1]
```diff
@@ -332,6 +332,86 @@ def collect_info(self, info):
...
+
+    Parameters
+    ----------
+    col_name : str
+        The name of the column that contains the feature.
+    feat_name : str
+        The feature name used in the constructed graph.
+    bucket_cnt: num:
+        The count of bucket lists used in the bucket feature transform
+    range: list[num]:
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:346]: bucket_range

File: python/graphstorm/gconstruct/transform.py [hunk 2]
```diff
@@ -332,6 +332,86 @@ def collect_info(self, info):
...
+        bucket_size = (max_val - min_val) / self.bucket_cnt
+        for i, f in enumerate(feats):
+            high_val = min(f + (self.slide_window_size / 2), max_val)
+            low_val = max(f - (self.slide_window_size / 2), min_val)
+
+            # Early exits to avoid numpy calls
+            membership_list = [0.0] * self.bucket_cnt
+            if f >= max_val:
+                membership_list[-1] = 1.0
+                encoding[i] = membership_list
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/gconstruct/transform.py:395]: BTW, if slide window is large enough, f may still cover more than 1 bucket.

File: python/graphstorm/gconstruct/transform.py [hunk 3]
```diff
@@ -332,6 +332,86 @@ def collect_info(self, info):
...
+            # Determine upper and lower bucket membership
+            low_val -= min_val
+            high_val -= min_val
+            low_idx = max(low_val // bucket_size, 0)
+            high_idx = min(high_val // bucket_size + 1, self.bucket_cnt)
+
+            idx = np.arange(start=low_idx, stop=high_idx, dtype=int)
+            membership_list = np.zeros(self.bucket_cnt, dtype=float)
+            membership_list[idx] = 1.0
+            encoding[i] = membership_list
```
Comment (classicsong) [inline python/graphstorm/gconstruct/transform.py:411]: Why not encoding[i][idx] = 1.0?

File: tests/unit-tests/gconstruct/test_transform.py [hunk 4]
```diff
@@ -633,6 +634,50 @@ def test_classification_processor():
...
+    else:
+        assert bucket_feats['test'].dtype == np.float32
+
+    feats_tar = np.array([[1, 0], [1, 0], [0, 1], [0, 1]], dtype=out_dtype)
+    assert_equal(bucket_feats['test'], feats_tar)
+
+    bucket_range = [10, 30]
+    transform = BucketTransform("test", "test", 2,
+                 bucket_range=bucket_range, slide_window_size=10, out_dtype=out_dtype)
+    feats = np.zeros(4)
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_transform.py:670]:
```suggestion
    feats = np.array([1,11,21,31)]
```

File: docs/source/configuration/configuration-gconstruction.rst [hunk 5]
```diff
@@ -90,6 +90,7 @@ Currently, the graph construction pipeline supports the following feature transf
...
...ge`` and ``bucket_cnt`` field, which ``range`` defines a numerical range, and ``bucket_cnt`` defines number of buckets among the range. All buckets will have same length, and each of them is left included. e.g, bucket ``(a, b)`` will include a, but not b. All input feature column data are categorized into respective buckets using this method. Any input data lower than the minimum value will be assigned to the first bucket, and any input data exceeding the maximum value will be assigned to the last bucket. For example, with range=`[10,30]` and bucket_cnt=`2`, input data `1` will fall into the bucket `[10, 20]`, input data `11` will be mapped to `[10, 20]`, input data `21` will be mapped to `[20, 30]`, input data `31` will be mapped to `[20, 30]`. Finally we use one-hot-encoding to encode the feature for each numerical bucket. If user wants to make numeric values fall into more than one bucket, it is preferred to use the `slide_window_size`: `"slide_window_size": s` , where `s` is a number. Then each value `v` will be transformed into a range from `v - s/2` through `v + s/2` , and assigns the value `v` to every bucket that the range covers, and do the one-hot-encoding just as before.
```
Comment (classicsong) [inline docs/source/configuration/configuration-gconstruction.rst:93]: `If user wants to make numeric` -> `If a user wants to make numeric`

File: tests/unit-tests/gconstruct/test_transform.py [hunk 6]
```diff
@@ -633,6 +634,79 @@ def test_classification_processor():
...
+    if out_dtype is not None:
+        assert bucket_feats['test'].dtype == np.float16
+    else:
+        assert bucket_feats['test'].dtype == np.float32
+
+    feats_tar = np.array([[1, 0], [1, 0], [0, 1], [0, 1]], dtype=out_dtype)
+    assert_equal(bucket_feats['test'], feats_tar)
+
+    # Case for float number
+    bucket_range = [1.1, 2.1, 3.1]
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_transform.py:654]: Why there are three numbers for a bucket range?

File: tests/unit-tests/gconstruct/test_transform.py [hunk 7]
```diff
@@ -633,6 +634,79 @@ def test_classification_processor():
...
+        assert bucket_feats['test'].dtype == np.float16
+    else:
+        assert bucket_feats['test'].dtype == np.float32
+
+    feats_tar = np.array([[1, 0], [1, 0], [0, 1], [0, 1]], dtype=out_dtype)
+    assert_equal(bucket_feats['test'], feats_tar)
+
+    # Edge case for large sliding window
+    bucket_range = [10, 30]
+    transform = BucketTransform("test", "test", 2,
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_transform.py:697]: Can you change num buckets to 2 and slide window to 20?

File: tests/unit-tests/gconstruct/test_transform.py (hunk 6 above)
Comment (jalencato) [reply tests/unit-tests/gconstruct/test_transform.py:654]: Test if it is provided with bucket range containing more elements, if it will fetch the maximum and minimum value

File: tests/unit-tests/gconstruct/test_transform.py (hunk 7 above)
Comment (classicsong) [reply tests/unit-tests/gconstruct/test_transform.py:697]: num buckets to 3

File: tests/unit-tests/gconstruct/test_transform.py (hunk 6 above)
Comment (jalencato) [reply tests/unit-tests/gconstruct/test_transform.py]: Already change that to two value, and add assert about that.

**Review by classicsong — APPROVED** [review]
LGTM
